core/src/test/java/com/google/instrumentation/stats/StatsContextFactoryTest.java

```diff
@@ -72,41 +73,43 @@ private static StatsContext testDeserialize(InputStream inputStream) throws IOEx
    * remove this method and use StatsContext.serialize() instead.
    * Currently StatsContext.serialize() can only serialize strings.
    */
-  private InputStream constructInputStream(int valueType) {
-    // TODO(songya): please note that ByteBuffer will overflow if > 1024 bytes were appended.
-    ByteBuffer buffer = ByteBuffer.allocate(1024);
-    VarInt.putVarInt(valueType, buffer);
-    encodeString(KEY1, buffer);
+  private static InputStream constructInputStream(int valueType) throws IOException {
+    ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+    VarInt.putVarInt(valueType, byteArrayOutputStream);
+    encodeString(KEY1, byteArrayOutputStream);
     switch (valueType) {
       case VALUE_TYPE_STRING:
-        encodeString(VALUE_STRING, buffer);
+        encodeString(VALUE_STRING, byteArrayOutputStream);
         break;
       case VALUE_TYPE_INTEGER:
-        encodeInteger(VALUE_INT, buffer);
+        encodeInteger(VALUE_INT, byteArrayOutputStream);
         break;
       case VALUE_TYPE_BOOLEAN:
-        encodeBoolean(VALUE_BOOL, buffer);
+        encodeBoolean(VALUE_BOOL, byteArrayOutputStream);
         break;
       default:
         return null;
     }
-    return new ByteArrayInputStream(buffer.array());
+    return new ByteArrayInputStream(byteArrayOutputStream.toByteArray());
   }
 
-  private static void encodeString(String input, ByteBuffer buffer) {
-    VarInt.putVarInt(input.length(), buffer);
+  private static void encodeString(String input, ByteArrayOutputStream byteArrayOutputStream)
+      throws IOException {
+    VarInt.putVarInt(input.length(), byteArrayOutputStream);
     for (int i = 0; i < input.length(); i++) {
-      buffer.put((byte) input.charAt(i));
+      byteArrayOutputStream.write((byte) input.charAt(i));
```

Comment:
Updated encoding in UTF-8 and add a few tests: https://github.com/google/instrumentation-java/pull/124/commits/64d8f37c0fc6d024ae9a8b6330abf75c5c432e1c